lib/Dialect/FIRRTL/Transforms/ExtractClasses.cpp

```diff
@@ -0,0 +1,146 @@
+/// properties.
+void ExtractClassesPass::extractClass(FModuleOp moduleOp) {
+  // Map from Values in the FModuleOp to Values in the ClassOp.
+  IRMapping mapping;
+
+  // Remember ports and operations to clean up when done.
+  llvm::BitVector portsToErase(moduleOp.getNumPorts());
+  SmallVector<Operation *> opsToErase;
+
+  // Collect information about input and output properties. Mark property ports
+  // to be erased.
+  SmallVector<Property> inputProperties;
+  SmallVector<Property> outputProperties;
+  for (auto [index, port] : llvm::enumerate(moduleOp.getPorts())) {
+    if (!isa<PropertyType>(port.type))
+      continue;
+
+    portsToErase.set(index);
+
+    if (port.isInput())
+      inputProperties.push_back({index, port.name, port.type, port.loc});
+
+    if (port.isOutput())
+      outputProperties.push_back({index, port.name, port.type, port.loc});
+  }
+
+  // If the FModuleOp has no properties, nothing to do.
+  if (inputProperties.empty() && outputProperties.empty())
+    return;
+
+  OpBuilder builder = OpBuilder::atBlockEnd(getOperation().getBody(0));
+
+  // Collect the parameter names from input properties.
+  SmallVector<StringRef> formalParamNames;
+  for (auto inputProperty : inputProperties)
+    formalParamNames.push_back(inputProperty.name);
+
+  // Construct the ClassOp with the FModuleOp name and parameter names.
+  auto classOp = builder.create<ClassOp>(moduleOp.getLoc(), moduleOp.getName(),
+                                         formalParamNames);
+
+  // Construct the ClassOp body with block arguments for each input property,
+  // updating the mapping to map from the input property to the block argument.
+  Block *classBody = &classOp.getRegion().emplaceBlock();
+  for (auto inputProperty : inputProperties) {
+    BlockArgument parameterValue =
+        classBody->addArgument(inputProperty.type, inputProperty.loc);
+    BlockArgument inputValue = moduleOp.getArgument(inputProperty.index);
+    mapping.map(inputValue, parameterValue);
+  }
+
+  // Construct ClassFieldOps for each output property.
+  builder.setInsertionPointToStart(classBody);
+  for (auto outputProperty : outputProperties) {
+    // Get the Value driven to the property to use for this ClassFieldOp.
+    auto outputValue =
+        cast<FIRRTLPropertyValue>(moduleOp.getArgument(outputProperty.index));
+    Value originalValue = getDriverFromConnect(outputValue);
+
+    // Mark the property assign to be erased.
+    opsToErase.push_back(getPropertyAssignment(outputValue));
+
+    // If the Value is defined by an Operation, copy that into the body, and
+    // map from the old Value to the new Value. This may need to walk property
+    // ops in order to copy them into the ClassOp, but for now only constant
+    // ops exist. Mark the property op to be erased.
+    if (auto *op = originalValue.getDefiningOp()) {
+      builder.clone(*op, mapping);
+      opsToErase.push_back(op);
```

Comment:
Perhaps is it necessary to make opsToErase a set if `originalValue` is multiply used?

Reply:
Good point, I'll add it to the tests and fix this. The order does matter here, so I guess SetVector.

Reply:
This can actually be done in a way where order doesn't actually matter (yet), so I'll use a simple set.

Reply:
Actually I added some logic to not duplicate the defining op here when there are multiple uses, so I think we can go back to just a list.

Reply:
Looks great!